AO3-7214 AO3-7215 Bookmark Javascript Improvement

ReverM opened the pull request:
## Issue

https://otwarchive.atlassian.net/browse/AO3-7214
https://otwarchive.atlassian.net/browse/AO3-7215

## Purpose

This adresses both tickets 7214 and 7215 by making all buttons that could edit a bookmark disappear when the form is open, by allowing a closed form to be reopened (7215) and by fixing the opening and closing of multiple forms at the same time.
7214 was fixed by removing the variables and jquerying the appropriate component each time the form is closed or reopened.
7215 was fixed by changing the href of the opening buttons as part of the closing event, as well as changing the id to a class so every element is affected at the same time.


## Credit

Danaël / Rever ( they / he )

Jira account goes by Danaël Villeneuve

ReverM committed:
First batch of change

Changed id to class so that every button that can do one feature behaves the same and changed the organisation of the jquery

ReverM committed:
Merge branch 'otwcode:master' into AO3-7215

ReverM committed:
Added test and translation

ReverM committed:
Added tests

ReverM committed:
Adressed style issues

ReverM committed:
Fixed translation

ReverM committed:
Remove leftover test

Since the new tests have been moved to their new feature file to test javascript feature these are not needed

ReverM committed:
These spaces thought we wouldn't see them

ReverM commented:
Appologies if having multiple ticket in one PR is bad etiquette. I decided to put both in the same PR, as they were extremely closely related, and would likely be waiting on each other if done separately.

ReverM committed:
Normalizing

sarken reviewed (changes requested):
Thanks for working on this! In addition to the question I left on the JavaScript file, I think you also need to change`id` to `class` for the buttons in the [bookmarkable_blurb](https://github.com/otwcode/otwarchive/blob/027cbf272402ca467a4c47cb7e4f0681c8342abc/app/views/bookmarks/_bookmarkable_blurb.html.erb#L16-L30), which is used for pages where we show multiple bookmarks under a single item, e.g., a tag's bookmark listing.

ReverM commented on app/views/bookmarks/bookmark_form_dynamic.js.erb:
```diff
@@ -2,21 +2,19 @@
+$j('#bookmark_form_close_for_<%= bookmark_form_id %>').click(function(){
+  $j('#bookmark_form_placement_for_<%= bookmark_form_id %>').hide();
+  $j('.bookmark_form_trigger_for_<%= bookmark_form_id %>').show();
+  $j('.bookmark_form_trigger_for_<%= bookmark_form_id %>').attr('href', '#');
```
Which line in particular are you referring to?

sarken replied:
>Which line in particular are you referring to?

The line removing the href, sorry! GitHub told me the comment was only on line 12 before I submitted the review, but it appears to have lied.

>If we don't do that, 7215 is reintroduced (since it will go through the controller to find the route of the link instead of the javascript watcher).

For me, 7215 remained fixed without this line. The only difference in behavior was the information I had entered in the form was lost upon re-opening it (which would be consistent with opening a new form instead of un-hiding the original). But based on your explanation, I suspect if I had re-opened the form twice instead of just once, I would've run into 7215. 

>An alternate solution would be to delete the form instead of hiding it.

That would mean losing any information you had entered in the form before closing it, so we don't want to do that. While we _could_ store that information somewhere and retrieve it when you re-open the form, it's probably not worth the extra code. If we get complaints about the removal of the right-click functionality, we can reconsider.

>Is the behavior you mentioned not already occurring if you close and open a bookmark on the works page having written stuff already?

The form currently (i.e., on production) keeps what I've written.

ReverM replied:
I believe if you go on the works index, open a bookmark form and close it again you can't open it with a right click anymore.

ReverM committed:
Changed an id to class

ReverM committed:
fixed linter issue

ReverM committed:
Normalized

ReverM committed:
Merge branch 'master' into AO3-7215

ReverM committed:
Added fix to allow right click

ReverM committed:
Merge branch 'AO3-7215' of https://github.com/ReverM/otwarchive into AO3-7215

ReverM committed:
Pushed gemfile.lock by accident

sarken commented on app/views/bookmarks/bookmark_form_dynamic.js.erb:
```diff
@@ -2,21 +2,19 @@
 
-bookmark_div.html("<%= escape_javascript(render "bookmarks/bookmark_form", :bookmarkable => @bookmarkable, :bookmark => @bookmark, :button_name => @button_name, :action => @action, :in_page => true, :dynamic => true) %>");
-bookmark_open.hide();
+if (!bookmark_div.children().length) { bookmark_div.html("<%= escape_javascript(render "bookmarks/bookmark_form", :bookmarkable => @bookmarkable, :bookmark => @bookmark, :button_name => @button_name, :action => @action, :in_page => true, :dynamic => true) %>"); }
```
I didn't check -- does this work with the new `bookmarkable: @bookmarkable` syntax? If so, it'd be nice to fix that while we're touching this line.

sarken commented on app/views/bookmarks/bookmark_form_dynamic.js.erb:
```diff
@@ -2,21 +2,19 @@
-bookmark_div.html("<%= escape_javascript(render "bookmarks/bookmark_form", :bookmarkable => @bookmarkable, :bookmark => @bookmark, :button_name => @button_name, :action => @action, :in_page => true, :dynamic => true) %>");
-bookmark_open.hide();
+if (!bookmark_div.children().length) { bookmark_div.html("<%= escape_javascript(render "bookmarks/bookmark_form", :bookmarkable => @bookmarkable, :bookmark => @bookmark, :button_name => @button_name, :action => @action, :in_page => true, :dynamic => true) %>"); }
+$j('.bookmark_form_trigger_for_<%= bookmark_form_id %>').hide();
```
Could we switch this and the other lines we're touching to double quotes?

sarken commented on config/locales/views/en.yml:
```diff
@@ -536,15 +536,34 @@ en:
+      navigation:
+        bookmarks: Bookmarks
+        edit: Edit Bookmark
+        new: Boorkmark
```
Typo -- there's an "r" before the "k" (ngl, I would not have noticed this without the tests!)

sarken commented on app/views/bookmarks/_bookmarkable_blurb.html.erb:
```diff
@@ -16,15 +16,9 @@
-                    edit_bookmark_path(current_user_bookmark),
-                    id: "bookmark_form_trigger_for_#{bookmarkable.id}",
-                    remote: true %>
+        <%= link_to t(".saved"), edit_bookmark_path(current_user_bookmark), data: { path: edit_bookmark_path(current_user_bookmark) }, class: "bookmark_form_trigger_for_#{bookmarkable.id}", remote: true %>
```
I think the multiline format was a little easier to read, if you could keep that here. (Reviewdog might have some opinions about the indenting, though. It usually takes me a few tries to figure out exactly what it wants.)

ReverM committed:
Adding more test coverage and adressing various PR comments

ReverM committed:
Addressing style and adding forgotten file

ReverM committed:
Added test for the rest of the buttons

ReverM commented:
I shall put this as ready for review as I do not understand why the step fails in the test.

ReverM commented on app/views/bookmarks/bookmark_form_dynamic.js.erb:
```diff
@@ -2,21 +2,19 @@
+  $j('.bookmark_form_trigger_for_<%= bookmark_form_id %>').hide();
   event.preventDefault();
-});
+});
```
Done

ReverM committed:
Changed to new syntax

sarken reviewed (changes requested):
Our JavaScript tests do tend to be flaky, unfortunately. If these continue to be a problem, we can update our [list of known test failures](https://github.com/otwcode/otwarchive/wiki/Automated-Testing#user-content-Known_Test_Failures) to mention `features/bookmarks` when they're merged.

ReverM committed:
Adressed comment and added new step for bookmark

Given how consistently the I bookmark the work X step seemed to be, I decided to add a new step definition. I could combine it with the previous one if that is simpler (as they are quite similar)

sarken commented on app/views/bookmarks/_bookmarkable_blurb.html.erb:
```diff
@@ -16,15 +16,16 @@
-                    id: "bookmark_form_trigger_for_#{bookmarkable.id}",
-                    remote: true %>
+        <%= link_to t(".saved"),
+        edit_bookmark_path(current_user_bookmark),
```
Can you address the Reviewdog comments here? I can never remember if it wants one level of indentation from the `<%=` or from `link_to`, but one of those options should make it stop complaining.

sarken reviewed (changes requested):
Just a few little things left, most of which are from Reviewdog! Thank you so much for working on this; I'm sorry it turned out to be so much more complicated than it looked.

ReverM commented on features/bookmarks/bookmark_javascript.feature:
```diff
@@ -0,0 +1,141 @@
+Feature: Edit bookmarks with javascript enabled
+  In order to have a good user experience
+  As a humble user
+  I want to be able to edit bookmarks with javascript enabled
```
What step exactly fails?

ReverM committed:
Requested changes

ReverM committed:
Is that what the dog wants?

ReverM committed:
The dog wanted more

ReverM committed:
Changed the header of the test

ReverM committed:
oop

sarken reviewed (approved):
🎉

ReverM committed:
Merge branch 'master' into AO3-7215
